# Data Preparation

Current file is dedicated to upload and explore the provided datasets. After exploration, the data will be cleaned and filtered according to the project needs. The resulted data will be stored as a new dataset in "Processed" folder.

# File Content

* **Import Packeges**
* **Upload Datasets**
* **Explore Data**
* **Clean Datasets**
* **Prepare Datasets**
* **Merge Datasets**
* **Store New DataFrame**

# Import Packages

Import necessary packages to access dataset and manipulate it. 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

from functools import reduce

# Import additional files with statistical functions
import sys
import os

module_path = os.path.abspath(os.path.join('../../src'))
if module_path not in sys.path:
    sys.path.append(module_path)
    
import explore_data as ed 

Adjust settings for the notebook view and set style preferences.

In [2]:
pd.options.display.max_rows = 500
pd.options.display.max_columns = 100
plt.style.use('seaborn')

# Upload Data

First, I upload Datasets into following variables:

- **rp_sale:** Real Property Sale Record
- **parcel:** Parcel Record
- **res_bl:** Residential Building Record

**\* All Datasets are uploaded as string variables in order to keep leading zeros**

In [3]:
rp_sale = pd.read_csv("../../data/raw/EXTR_RPSale.csv", dtype=str)
parcel = pd.read_csv("../../data/raw/EXTR_Parcel.csv", dtype = str)
res_bl = pd.read_csv ("../../data/raw/EXTR_ResBldg.csv", dtype=str)

# Explore Datasets

Now, explore Data and check following information:
* What columns do we have in each of the datasets
* Are there any missing values in tables
* Are there duplicates in data

In [4]:
ed.show_info(rp_sale)

Lenght of Dataset: 351067
                    missing_values_% Data_type  Duplicates
ExciseTaxNbr                     0.0    object           0
Major                            0.0    object           0
Minor                            0.0    object           0
DocumentDate                     0.0    object           0
SalePrice                        0.0    object           0
RecordingNbr                     0.0    object           0
Volume                           0.0    object           0
Page                             0.0    object           0
PlatNbr                          0.0    object           0
PlatType                         0.0    object           0
PlatLot                          0.0    object           0
PlatBlock                        0.0    object           0
SellerName                       0.0    object           0
BuyerName                        0.0    object           0
PropertyType                     0.0    object           0
PrincipalUse                  

In [5]:
ed.show_info(parcel)

Lenght of Dataset: 205199
                        missing_values_% Data_type  Duplicates
Unnamed: 0                      0.000000    object           0
Major                           0.000000    object           0
Minor                           0.000000    object           0
PropName                        4.440080    object           0
PlatName                       13.910887    object           0
PlatLot                         0.000000    object           0
PlatBlock                       0.000000    object           0
Range                           0.000000    object           0
Township                        0.000000    object           0
Section                         0.000000    object           0
QuarterSection                  0.000000    object           0
PropType                        0.000000    object           0
Area                            0.002924    object           0
SubArea                         0.002924    object           0
SpecArea                     

In [6]:
ed.show_info(res_bl)

Lenght of Dataset: 181510
                    missing_values_% Data_type  Duplicates
Major                       0.000000    object           0
Minor                       0.000000    object           0
BldgNbr                     0.000000    object           0
NbrLivingUnits              0.000000    object           0
Address                     0.000000    object           0
BuildingNumber              0.000000    object           0
Fraction                    0.000000    object           0
DirectionPrefix             0.200540    object           0
StreetName                  0.000000    object           0
StreetType                  0.000000    object           0
DirectionSuffix             0.200540    object           0
ZipCode                    14.828935    object           0
Stories                     0.000000    object           0
BldgGrade                   0.000000    object           0
BldgGradeVar                0.000000    object           0
SqFt1stFloor                0.

**Eploration Results:** In parcel dataset the columns PlatName, SpecArea and SpecSubArea have missing values. The rs_bl dataset has missing values in ZipCode column. 

# Data Cleaning

To be able to manipulate data and merge three given datasets, I will convert necessary columns to numeric data types and clean datasets from unnecessary columns.

## Drop Columns

From the "rp_sale" dataset I will drop needless columns that are categorical. In order to find those columns I used the "Data Dictionary" file in references folder. 

**RP_SALE DATASET**

From this dataset I will drop columns that are irrelevant to project needs.

In [7]:
rp_sale.drop(columns = ['RecordingNbr', 'Volume', 'Page', 
            'PlatNbr', 'PlatType', 'PlatLot', 'PlatBlock', 
            'SellerName', 'BuyerName', 'AFForestLand', 'AFCurrentUseLand', 
            'AFNonProfitUse','AFHistoricProperty','SaleWarning'], inplace = True)
rp_sale.columns

Index(['ExciseTaxNbr', 'Major', 'Minor', 'DocumentDate', 'SalePrice',
       'PropertyType', 'PrincipalUse', 'SaleInstrument', 'SaleReason',
       'PropertyClass'],
      dtype='object')

**PARCEL DATASET**

The parcel data has a lot of columns that are irrelevant to sale price, thus will be dropped.

In [8]:
parcel.drop(columns = ['Unnamed: 0', 'Range', 'Township', 'Section', 'PropName', 'PlatName', 'PlatLot',
       'PlatBlock', 'QuarterSection','Area', 'SubArea', 'SpecArea', 'SpecSubArea',
       'DistrictName', 'LevyCode', 'CurrentZoning',
       'InadequateParking', 'PcntUnusable', 'Unbuildable', 'MtRainier',
       'Olympics', 'WfntAccessRights', 'WfntProximityInfluence',
       'PowerLines', 'OtherNuisances', 
       'DNRLease', 'AdjacentGolfFairway', 'AdjacentGreenbelt', 'NativeGrowthProtEsmt', 'Easements',
       'OtherDesignation', 'DeedRestrictions', 'DevelopmentRightsPurch',
       'CoalMineHazard', 'CriticalDrainage', 'ErosionHazard', 'LandfillBuffer',
       'HundredYrFloodPlain', 'SeismicHazard', 'LandslideHazard',
       'SteepSlopeHazard', 'Stream', 'Wetland', 'SpeciesOfConcern',
       'SensitiveAreaTract', 'WaterProblems', 'TranspConcurrency',
       'OtherProblems', 'AirportNoise', 'Cascades', 'Territorial', 'SeattleSkyline',
       'PugetSound', 'LakeWashington', 'LakeSammamish', 'SmallLakeRiverCreek',
       'OtherView', 'HistoricSite', 'TrafficNoise', 'WfntRestrictedAccess'], inplace = True)
parcel.columns

Index(['Major', 'Minor', 'PropType', 'HBUAsIfVacant', 'HBUAsImproved',
       'PresentUse', 'SqFtLot', 'WaterSystem', 'SewerSystem', 'Access',
       'Topography', 'StreetSurface', 'RestrictiveSzShape', 'WfntLocation',
       'WfntFootage', 'WfntBank', 'WfntPoorQuality', 'TidelandShoreland',
       'LotDepthFactor', 'NbrBldgSites', 'Contamination',
       'CurrentUseDesignation'],
      dtype='object')

**RES_BL DATASET**

In [9]:
res_bl.drop(columns = ['NbrLivingUnits', 'BldgNbr', 'Address',
       'BuildingNumber', 'Fraction', 'DirectionPrefix', 'StreetName', 'DaylightBasement',
       'StreetType', 'DirectionSuffix', 'ZipCode', 'Stories', 'ViewUtilization', 'BldgGradeVar'],inplace = True)
res_bl.columns

Index(['Major', 'Minor', 'BldgGrade', 'SqFt1stFloor', 'SqFtHalfFloor',
       'SqFt2ndFloor', 'SqFtUpperFloor', 'SqFtUnfinFull', 'SqFtUnfinHalf',
       'SqFtTotLiving', 'SqFtTotBasement', 'SqFtFinBasement',
       'FinBasementGrade', 'SqFtGarageBasement', 'SqFtGarageAttached',
       'SqFtOpenPorch', 'SqFtEnclosedPorch', 'SqFtDeck', 'HeatSystem',
       'HeatSource', 'BrickStone', 'Bedrooms', 'BathHalfCount',
       'Bath3qtrCount', 'BathFullCount', 'FpSingleStory', 'FpMultiStory',
       'FpFreestanding', 'FpAdditional', 'YrBuilt', 'YrRenovated',
       'PcntComplete', 'Obsolescence', 'PcntNetCondition', 'Condition',
       'AddnlCost'],
      dtype='object')

**Column Dropping Results:** The rp_sale dataset has 10 columns, the parcel dataset has 22 columns, and res_bl has 36 columns.

## Change Data Type

Using the "Data Dictionary" in references folder, I convert necessary columns into numeric data types. 

**RP_SALE DATASET**

In [10]:
rp_sale['ExciseTaxNbr'] = rp_sale.ExciseTaxNbr.astype(int)
rp_sale['SalePrice'] = rp_sale.SalePrice.astype(float)
rp_sale['PropertyType'] = rp_sale.PropertyType.astype(int)
rp_sale['PrincipalUse'] = rp_sale.PrincipalUse.astype(int)
rp_sale['SaleInstrument'] = rp_sale.SaleInstrument.astype(int)
rp_sale['SaleReason'] = rp_sale.SaleReason.astype(int)
rp_sale['PropertyClass'] = rp_sale.PropertyClass.astype(int)
rp_sale['DocumentDate'] = pd.to_datetime(rp_sale['DocumentDate'])


Check the result:

In [11]:
ed.show_info(rp_sale)

Lenght of Dataset: 351067
                missing_values_%       Data_type  Duplicates
ExciseTaxNbr                 0.0           int64           0
Major                        0.0          object           0
Minor                        0.0          object           0
DocumentDate                 0.0  datetime64[ns]           0
SalePrice                    0.0         float64           0
PropertyType                 0.0           int64           0
PrincipalUse                 0.0           int64           0
SaleInstrument               0.0           int64           0
SaleReason                   0.0           int64           0
PropertyClass                0.0           int64           0


**PARCEL DATASET**

In [12]:
parcel['HBUAsIfVacant'] = parcel.HBUAsIfVacant.astype(int)
parcel['HBUAsImproved'] = parcel.HBUAsImproved.astype(int)
parcel['PresentUse'] = parcel.PresentUse.astype(int)
parcel['SqFtLot'] = parcel.SqFtLot.astype(int)
parcel['WaterSystem'] = parcel.WaterSystem.astype(int)
parcel['SewerSystem'] = parcel.SewerSystem.astype(int)
parcel['Access'] = parcel.Access.astype(int)
parcel['Topography'] = parcel.Topography.astype(int)
parcel['StreetSurface'] = parcel.StreetSurface.astype(int)
parcel['RestrictiveSzShape'] = parcel.RestrictiveSzShape.astype(int)
parcel['WfntLocation'] = parcel.WfntLocation.astype(int)
parcel['WfntFootage'] = parcel.WfntFootage.astype(int)
parcel['WfntBank'] = parcel.WfntBank.astype(int)
parcel['WfntPoorQuality'] = parcel.WfntPoorQuality.astype(int)
parcel['TidelandShoreland'] = parcel.TidelandShoreland.astype(int)
parcel['LotDepthFactor'] = parcel.LotDepthFactor.astype(int)
parcel['NbrBldgSites'] = parcel.NbrBldgSites.astype(int)
parcel['Contamination'] = parcel.Contamination.astype(int)
parcel['CurrentUseDesignation'] = parcel.CurrentUseDesignation.astype(int)


Check the result:

In [13]:
ed.show_info(parcel)

Lenght of Dataset: 205199
                       missing_values_% Data_type  Duplicates
Major                               0.0    object           0
Minor                               0.0    object           0
PropType                            0.0    object           0
HBUAsIfVacant                       0.0     int64           0
HBUAsImproved                       0.0     int64           0
PresentUse                          0.0     int64           0
SqFtLot                             0.0     int64           0
WaterSystem                         0.0     int64           0
SewerSystem                         0.0     int64           0
Access                              0.0     int64           0
Topography                          0.0     int64           0
StreetSurface                       0.0     int64           0
RestrictiveSzShape                  0.0     int64           0
WfntLocation                        0.0     int64           0
WfntFootage                         0.0     

**RES_BL DATASET**

In [14]:
res_bl['BldgGrade'] = res_bl.BldgGrade.astype(int)
res_bl['SqFt1stFloor'] = res_bl.SqFt1stFloor.astype(int)
res_bl['SqFtHalfFloor'] = res_bl.SqFtHalfFloor.astype(int)
res_bl['SqFt2ndFloor'] = res_bl.SqFt2ndFloor.astype(int)
res_bl['SqFtUpperFloor'] = res_bl.SqFtUpperFloor.astype(int)
res_bl['SqFtUnfinFull'] = res_bl.SqFtUnfinFull.astype(int)
res_bl['SqFtUnfinHalf'] = res_bl.SqFtUnfinHalf.astype(int)
res_bl['SqFtTotLiving'] = res_bl.SqFtTotLiving.astype(int)
res_bl['SqFtTotBasement'] = res_bl.SqFtTotBasement.astype(int)
res_bl['SqFtFinBasement'] = res_bl.SqFtFinBasement.astype(int)
res_bl['FinBasementGrade'] = res_bl.FinBasementGrade.astype(int)
res_bl['SqFtGarageBasement'] = res_bl.SqFtGarageBasement.astype(int)
res_bl['SqFtGarageAttached'] = res_bl.SqFtGarageAttached.astype(int)
res_bl['SqFtOpenPorch'] = res_bl.SqFtOpenPorch.astype(int)
res_bl['SqFtEnclosedPorch'] = res_bl.SqFtEnclosedPorch.astype(int)
res_bl['SqFtDeck'] = res_bl.SqFtDeck.astype(int)
res_bl['HeatSystem'] = res_bl.HeatSystem.astype(int)
res_bl['HeatSource'] = res_bl.HeatSource.astype(int)
res_bl['BrickStone'] = res_bl.BrickStone.astype(int)
res_bl['Bedrooms'] = res_bl.Bedrooms.astype(int)
res_bl['BathHalfCount'] = res_bl.BathHalfCount.astype(int)
res_bl['Bath3qtrCount'] = res_bl.Bath3qtrCount.astype(int)
res_bl['BathFullCount'] = res_bl.BathFullCount.astype(int)
res_bl['FpSingleStory'] = res_bl.FpSingleStory.astype(int)
res_bl['FpMultiStory'] = res_bl.FpMultiStory.astype(int)
res_bl['FpFreestanding'] = res_bl.FpFreestanding.astype(int)
res_bl['FpAdditional'] = res_bl.FpAdditional.astype(int)
res_bl['YrBuilt'] = res_bl.YrBuilt.astype(int)
res_bl['YrRenovated'] = res_bl.YrRenovated.astype(int)
res_bl['PcntComplete'] = res_bl.PcntComplete.astype(int)
res_bl['Obsolescence'] = res_bl.Obsolescence.astype(int)
res_bl['PcntNetCondition'] = res_bl.PcntNetCondition.astype(int)
res_bl['Condition'] = res_bl.Condition.astype(int)
res_bl['AddnlCost'] = res_bl.AddnlCost.astype(int)

Check the result:

In [15]:
ed.show_info(res_bl)

Lenght of Dataset: 181510
                    missing_values_% Data_type  Duplicates
Major                            0.0    object         191
Minor                            0.0    object         191
BldgGrade                        0.0     int64         191
SqFt1stFloor                     0.0     int64         191
SqFtHalfFloor                    0.0     int64         191
SqFt2ndFloor                     0.0     int64         191
SqFtUpperFloor                   0.0     int64         191
SqFtUnfinFull                    0.0     int64         191
SqFtUnfinHalf                    0.0     int64         191
SqFtTotLiving                    0.0     int64         191
SqFtTotBasement                  0.0     int64         191
SqFtFinBasement                  0.0     int64         191
FinBasementGrade                 0.0     int64         191
SqFtGarageBasement               0.0     int64         191
SqFtGarageAttached               0.0     int64         191
SqFtOpenPorch                 

**Convertion Results:** It is clear that now the res_bl dataframe has duplicates.

## Prepare Datasets

In order to merge datasets, I will concatinate "Major" and "Minor" columns and store the value in "PIN" column, since both that columns represent unique Identification Number for each sale

**RP_SALE DATASET**

In [16]:
rp_sale["PIN"] = rp_sale.Major + rp_sale.Minor
rp_sale.columns

Index(['ExciseTaxNbr', 'Major', 'Minor', 'DocumentDate', 'SalePrice',
       'PropertyType', 'PrincipalUse', 'SaleInstrument', 'SaleReason',
       'PropertyClass', 'PIN'],
      dtype='object')

Reorder the table, so the PIN column is first.

In [17]:
cols = list(rp_sale.columns)
cols = [cols[10]] + cols[:10]
rp_sale = rp_sale[cols]
rp_sale.head()

PIN  ExciseTaxNbr   Major Minor DocumentDate  SalePrice  \
0  1989201430       2857854  198920  1430   2017-03-28        0.0   
1  6385800110       2743355  638580  0110   2015-07-14   190000.0   
2  9197150200       2999169  919715  0200   2019-07-08   192000.0   
3  8946770240       2841697  894677  0240   2016-12-21   818161.0   
4  4458720260       2826129  445872  0260   2016-10-03        0.0   

   PropertyType  PrincipalUse  SaleInstrument  SaleReason  PropertyClass  
0             3             7              15          16              2  
1             3             6               3           1              8  
2             3             2               3           1              3  
3             2             6               3           1              8  
4             3             2              15          18              3

**PARCEL DATASET**

In [18]:
parcel["PIN"] = parcel.Major + parcel.Minor
parcel.columns

Index(['Major', 'Minor', 'PropType', 'HBUAsIfVacant', 'HBUAsImproved',
       'PresentUse', 'SqFtLot', 'WaterSystem', 'SewerSystem', 'Access',
       'Topography', 'StreetSurface', 'RestrictiveSzShape', 'WfntLocation',
       'WfntFootage', 'WfntBank', 'WfntPoorQuality', 'TidelandShoreland',
       'LotDepthFactor', 'NbrBldgSites', 'Contamination',
       'CurrentUseDesignation', 'PIN'],
      dtype='object')

Reorder the table, so the PIN column is first.

In [19]:
cols = list(parcel.columns)
cols = [cols[22]] + cols[:22]
parcel = parcel[cols]
parcel.head()

PIN   Major Minor PropType  HBUAsIfVacant  HBUAsImproved  \
0  8078410410  807841  0410        R              1              1   
1  7550800015  755080  0015        R              1              1   
2  8886000135  888600  0135        R              1              1   
3  0226039181  022603  9181        R              1              1   
4  2296700160  229670  0160        R              1              1   

   PresentUse  SqFtLot  WaterSystem  SewerSystem  Access  Topography  \
0           2     7424            2            2       4           0   
1           2     5000            2            2       4           0   
2           2   277041            2            1       4           0   
3           2    10560            2            2       4           0   
4           2     9853            2            2       4           0   

   StreetSurface  RestrictiveSzShape  WfntLocation  WfntFootage  WfntBank  \
0              1                   0             0            0         0   
1              1                   0             0            0         0   
2              1                   0             0            0         0   
3              1                   0             0            0         0   
4              1                   0             0            0         0   

   WfntPoorQuality  TidelandShoreland  LotDepthFactor  NbrBldgSites  \
0                0                  0               0             0   
1                0                  0               0             0   
2                0                  0               0             0   
3                0                  0               0             0   
4                0                  0               0             0   

   Contamination  CurrentUseDesignation  
0              0                      0  
1              0                      0  
2              0                      0  
3              0                      0  
4              0                      0

**RES_BL DATASET**

In [20]:
res_bl["PIN"] = res_bl.Major + res_bl.Minor
res_bl.columns

Index(['Major', 'Minor', 'BldgGrade', 'SqFt1stFloor', 'SqFtHalfFloor',
       'SqFt2ndFloor', 'SqFtUpperFloor', 'SqFtUnfinFull', 'SqFtUnfinHalf',
       'SqFtTotLiving', 'SqFtTotBasement', 'SqFtFinBasement',
       'FinBasementGrade', 'SqFtGarageBasement', 'SqFtGarageAttached',
       'SqFtOpenPorch', 'SqFtEnclosedPorch', 'SqFtDeck', 'HeatSystem',
       'HeatSource', 'BrickStone', 'Bedrooms', 'BathHalfCount',
       'Bath3qtrCount', 'BathFullCount', 'FpSingleStory', 'FpMultiStory',
       'FpFreestanding', 'FpAdditional', 'YrBuilt', 'YrRenovated',
       'PcntComplete', 'Obsolescence', 'PcntNetCondition', 'Condition',
       'AddnlCost', 'PIN'],
      dtype='object')

Reorder the table, so the PIN column is first.

In [21]:
cols = list(res_bl.columns)
cols = [cols[36]] + cols[:36]
res_bl = res_bl[cols]
res_bl.head()

PIN   Major Minor  BldgGrade  SqFt1stFloor  SqFtHalfFloor  \
0  0098000720  009800  0720         11          1970              0   
1  0098020140  009802  0140         10          1610              0   
2  0098300020  009830  0020         10          2520              0   
3  0098300160  009830  0160         10          2210              0   
4  0100500180  010050  0180          7           910              0   

   SqFt2ndFloor  SqFtUpperFloor  SqFtUnfinFull  SqFtUnfinHalf  SqFtTotLiving  \
0          2130               0              0              0           4100   
1          1400               0              0              0           3010   
2          2560               0              0              0           5080   
3          1860               0              0              0           4070   
4           700               0              0              0           1610   

   SqFtTotBasement  SqFtFinBasement  FinBasementGrade  SqFtGarageBasement  \
0                0                0                 0                   0   
1                0                0                 0                   0   
2                0                0                 0                   0   
3                0                0                 0                   0   
4                0                0                 0                   0   

   SqFtGarageAttached  SqFtOpenPorch  SqFtEnclosedPorch  SqFtDeck  HeatSystem  \
0                 750              0                  0         0           5   
1                 660            380                  0         0           5   
2                1020            360                  0       270           5   
3                1000            690                  0         0           5   
4                 440             60                  0         0           5   

   HeatSource  BrickStone  Bedrooms  BathHalfCount  Bath3qtrCount  \
0           2           0         4              1              0   
1           2           0         4              1              2   
2           2           0         4              0              0   
3           2           0         4              1              0   
4           2           0         3              1              1   

   BathFullCount  FpSingleStory  FpMultiStory  FpFreestanding  FpAdditional  \
0              3              3             0               0             0   
1              2              3             0               0             0   
2              5              1             0               0             0   
3              3              2             0               0             0   
4              1              1             0               0             0   

   YrBuilt  YrRenovated  PcntComplete  Obsolescence  PcntNetCondition  \
0     2001            0             0             0                 0   
1     2004            0             0             0                 0   
2     2017            0             0             0                 0   
3     2013            0             0             0                 0   
4     1994            0             0             0                 0   

   Condition  AddnlCost  
0          3          0  
1          3          0  
2          3          0  
3          3          0  
4          4          0

## Merge Datasets

To further investigate the provided data I will **merge all three dataframes into one.** This will help me to manipulate the data and build a model.

Make a list of dataframe names

In [22]:
data_sets = [parcel,rp_sale,res_bl]

Merge three datasets based on "PIN" columns

In [23]:
df_merged = reduce(lambda  left,right: pd.merge(left,right,on=['PIN'],
                                            how='outer'), data_sets)
df_merged.head()

PIN Major_x Minor_x PropType  HBUAsIfVacant  HBUAsImproved  \
0  8078410410  807841    0410        R            1.0            1.0   
1  7550800015  755080    0015        R            1.0            1.0   
2  8886000135  888600    0135        R            1.0            1.0   
3  0226039181  022603    9181        R            1.0            1.0   
4  2296700160  229670    0160        R            1.0            1.0   

   PresentUse   SqFtLot  WaterSystem  SewerSystem  Access  Topography  \
0         2.0    7424.0          2.0          2.0     4.0         0.0   
1         2.0    5000.0          2.0          2.0     4.0         0.0   
2         2.0  277041.0          2.0          1.0     4.0         0.0   
3         2.0   10560.0          2.0          2.0     4.0         0.0   
4         2.0    9853.0          2.0          2.0     4.0         0.0   

   StreetSurface  RestrictiveSzShape  WfntLocation  WfntFootage  WfntBank  \
0            1.0                 0.0           0.0          0.0       0.0   
1            1.0                 0.0           0.0          0.0       0.0   
2            1.0                 0.0           0.0          0.0       0.0   
3            1.0                 0.0           0.0          0.0       0.0   
4            1.0                 0.0           0.0          0.0       0.0   

   WfntPoorQuality  TidelandShoreland  LotDepthFactor  NbrBldgSites  \
0              0.0                0.0             0.0           0.0   
1              0.0                0.0             0.0           0.0   
2              0.0                0.0             0.0           0.0   
3              0.0                0.0             0.0           0.0   
4              0.0                0.0             0.0           0.0   

   Contamination  CurrentUseDesignation  ExciseTaxNbr Major_y Minor_y  \
0            0.0                    0.0       3005895  807841    0410   
1            0.0                    0.0       3015395  755080    0015   
2            0.0                    0.0       2838708  888600    0135   
3            0.0                    0.0       2743363  022603    9181   
4            0.0                    0.0       3002350  229670    0160   

  DocumentDate  SalePrice  PropertyType  PrincipalUse  SaleInstrument  \
0   2019-08-16   847888.0            11             6               3   
1   2019-10-12   819000.0            11             6               3   
2   2016-12-08        0.0             3             6              15   
3   2015-07-13   800000.0             3             6               3   
4   2019-07-21   730000.0             3             6               3   

   SaleReason  PropertyClass   Major Minor  BldgGrade  SqFt1stFloor  \
0           1              8  807841  0410        8.0        1360.0   
1           1              8  755080  0015        7.0         810.0   
2          13              8  888600  0135        8.0        1620.0   
3           1              8  022603  9181        9.0        2350.0   
4           1              8  229670  0160        7.0        1210.0   

   SqFtHalfFloor  SqFt2ndFloor  SqFtUpperFloor  SqFtUnfinFull  SqFtUnfinHalf  \
0            0.0         840.0             0.0            0.0            0.0   
1            0.0           0.0             0.0            0.0            0.0   
2            0.0        1340.0             0.0            0.0            0.0   
3            0.0           0.0             0.0            0.0            0.0   
4            0.0           0.0             0.0            0.0            0.0   

   SqFtTotLiving  SqFtTotBasement  SqFtFinBasement  FinBasementGrade  \
0         2200.0              0.0              0.0               0.0   
1         1800.0            990.0            990.0               7.0   
2         2960.0              0.0              0.0               0.0   
3         4200.0           2410.0           1850.0               8.0   
4         2410.0           1200.0           1200.0               6.0   

   SqFtGarageBasement  SqFtGa

In [24]:
ed.show_info(df_merged)

Lenght of Dataset: 354490
                       missing_values_%       Data_type  Duplicates
PIN                            0.000000          object         346
Major_x                       18.684871          object         346
Minor_x                       18.684871          object         346
PropType                      18.684871          object         346
HBUAsIfVacant                 18.684871         float64         346
HBUAsImproved                 18.684871         float64         346
PresentUse                    18.684871         float64         346
SqFtLot                       18.684871         float64         346
WaterSystem                   18.684871         float64         346
SewerSystem                   18.684871         float64         346
Access                        18.684871         float64         346
Topography                    18.684871         float64         346
StreetSurface                 18.684871         float64         346
RestrictiveSzShape    

**Merging Results:** Most of the columns have missing data and the number of duplicated values increased.

### Drop rows and columns

I will drop the rows with any missing value in order to clean up my dataframe.

In [25]:
df_merged.dropna(inplace = True)
ed.show_info(df_merged)

Lenght of Dataset: 251300
                       missing_values_%       Data_type  Duplicates
PIN                                 0.0          object         346
Major_x                             0.0          object         346
Minor_x                             0.0          object         346
PropType                            0.0          object         346
HBUAsIfVacant                       0.0         float64         346
HBUAsImproved                       0.0         float64         346
PresentUse                          0.0         float64         346
SqFtLot                             0.0         float64         346
WaterSystem                         0.0         float64         346
SewerSystem                         0.0         float64         346
Access                              0.0         float64         346
Topography                          0.0         float64         346
StreetSurface                       0.0         float64         346
RestrictiveSzShape    

I will drop the data from table that is irelevant to analysis. Filter out the needless Property types, Principal Use and Property Class. Keep only data that has "One family houselod" type, "Residential" Principal Use and "Residential Improved" property class.

In [26]:
index_ptype = df_merged[~(df_merged['PropertyType'] == 11)].index 
df_merged.drop(index_ptype, inplace=True)

index_puse = df_merged[~(df_merged['PrincipalUse'] == 6)].index
df_merged.drop(index_puse, inplace=True)

index_pclass = df_merged[~(df_merged['PropertyClass'] == 8)].index 
df_merged.drop(index_pclass, inplace=True)

index_pruse = df_merged[~(df_merged['PresentUse'] == 2)].index 
df_merged.drop(index_pruse, inplace=True)

index_ptype2 = df_merged[~(df_merged['PropType'] == "R")].index 
df_merged.drop(index_ptype2, inplace=True)

index_salep = df_merged[(df_merged['SalePrice'] == 0)].index 
df_merged.drop(index_salep, inplace=True)

In [27]:
ed.show_info(df_merged)

Lenght of Dataset: 61915
                       missing_values_%       Data_type  Duplicates
PIN                                 0.0          object           5
Major_x                             0.0          object           5
Minor_x                             0.0          object           5
PropType                            0.0          object           5
HBUAsIfVacant                       0.0         float64           5
HBUAsImproved                       0.0         float64           5
PresentUse                          0.0         float64           5
SqFtLot                             0.0         float64           5
WaterSystem                         0.0         float64           5
SewerSystem                         0.0         float64           5
Access                              0.0         float64           5
Topography                          0.0         float64           5
StreetSurface                       0.0         float64           5
RestrictiveSzShape     

In [28]:
df_merged.columns

Index(['PIN', 'Major_x', 'Minor_x', 'PropType', 'HBUAsIfVacant',
       'HBUAsImproved', 'PresentUse', 'SqFtLot', 'WaterSystem', 'SewerSystem',
       'Access', 'Topography', 'StreetSurface', 'RestrictiveSzShape',
       'WfntLocation', 'WfntFootage', 'WfntBank', 'WfntPoorQuality',
       'TidelandShoreland', 'LotDepthFactor', 'NbrBldgSites', 'Contamination',
       'CurrentUseDesignation', 'ExciseTaxNbr', 'Major_y', 'Minor_y',
       'DocumentDate', 'SalePrice', 'PropertyType', 'PrincipalUse',
       'SaleInstrument', 'SaleReason', 'PropertyClass', 'Major', 'Minor',
       'BldgGrade', 'SqFt1stFloor', 'SqFtHalfFloor', 'SqFt2ndFloor',
       'SqFtUpperFloor', 'SqFtUnfinFull', 'SqFtUnfinHalf', 'SqFtTotLiving',
       'SqFtTotBasement', 'SqFtFinBasement', 'FinBasementGrade',
       'SqFtGarageBasement', 'SqFtGarageAttached', 'SqFtOpenPorch',
       'SqFtEnclosedPorch', 'SqFtDeck', 'HeatSystem', 'HeatSource',
       'BrickStone', 'Bedrooms', 'BathHalfCount', 'Bath3qtrCount',
       'BathF

**Drop needless columns:**

In [29]:
df_merged.drop(columns = ['Major_y', 'Minor_y', 'Major', 'Minor', 'PropertyType',
                          'PrincipalUse', 'PropertyClass','PresentUse', 'WfntBank',
                          'WfntPoorQuality', 'TidelandShoreland', 'Contamination',
                          'PcntComplete'],inplace = True)
df_merged = df_merged.rename(columns = {'Major_x': 'Major', 'Minor_x': 'Minor'})

df_merged.columns

Index(['PIN', 'Major', 'Minor', 'PropType', 'HBUAsIfVacant', 'HBUAsImproved',
       'SqFtLot', 'WaterSystem', 'SewerSystem', 'Access', 'Topography',
       'StreetSurface', 'RestrictiveSzShape', 'WfntLocation', 'WfntFootage',
       'LotDepthFactor', 'NbrBldgSites', 'CurrentUseDesignation',
       'ExciseTaxNbr', 'DocumentDate', 'SalePrice', 'SaleInstrument',
       'SaleReason', 'BldgGrade', 'SqFt1stFloor', 'SqFtHalfFloor',
       'SqFt2ndFloor', 'SqFtUpperFloor', 'SqFtUnfinFull', 'SqFtUnfinHalf',
       'SqFtTotLiving', 'SqFtTotBasement', 'SqFtFinBasement',
       'FinBasementGrade', 'SqFtGarageBasement', 'SqFtGarageAttached',
       'SqFtOpenPorch', 'SqFtEnclosedPorch', 'SqFtDeck', 'HeatSystem',
       'HeatSource', 'BrickStone', 'Bedrooms', 'BathHalfCount',
       'Bath3qtrCount', 'BathFullCount', 'FpSingleStory', 'FpMultiStory',
       'FpFreestanding', 'FpAdditional', 'YrBuilt', 'YrRenovated',
       'Obsolescence', 'PcntNetCondition', 'Condition', 'AddnlCost'],
      dtype='obj

**Remove duplicates**

In [30]:
df_merged.drop_duplicates(subset=['PIN'], inplace = True)
ed.show_info(df_merged)

Lenght of Dataset: 57703
                       missing_values_%       Data_type  Duplicates
PIN                                 0.0          object           0
Major                               0.0          object           0
Minor                               0.0          object           0
PropType                            0.0          object           0
HBUAsIfVacant                       0.0         float64           0
HBUAsImproved                       0.0         float64           0
SqFtLot                             0.0         float64           0
WaterSystem                         0.0         float64           0
SewerSystem                         0.0         float64           0
Access                              0.0         float64           0
Topography                          0.0         float64           0
StreetSurface                       0.0         float64           0
RestrictiveSzShape                  0.0         float64           0
WfntLocation           

**Merging Results:** After additional cleaning the dataframe has total of 57703 observations, 56 columns and no missing values or duplicates.

### Store New DataFrame

Store the cleaned and merged table as "merged.csv" in Processed folder of Data directory. 

In [31]:
pd.DataFrame.to_csv(df_merged, '../../data/processed/merged.csv', sep=',', na_rep=' ', index=False)

**The further exploration and analysis of merged.csv file will be done in "data_preparation" notebook.**